crates/virtio-queue/src/lib.rs

```diff
@@ -190,4 +192,31 @@ pub trait QueueStateT: for<'a> QueueStateGuard<'a> {
 
     /// Set the index for the next descriptor in the used ring.
     fn set_next_used(&mut self, next_used: u16);
+
+    /// Pop and return the next available descriptor chain, or `None` when there are no more
+    /// descriptor chains available.
+    ///
+    /// This enables the consumption of available descriptor chains in a "one at a time"
+    /// manner, without having to hold a borrow after the method returns.
+    fn pop_descriptor_chain<M>(&mut self, mem: M) -> Option<DescriptorChain<M>>
+    where
+        M: Clone + Deref,
+        M::Target: GuestMemory;
+}
+
+/// Trait to access and manipulate a Virtio queue that's known to be exclusively accessed
+/// by a single execution thread.
+pub trait QueueStateOwnedT: QueueStateT {
+    /// Get a consuming iterator over all available descriptor chain heads offered by the driver.
+    ///
+    /// # Arguments
+    /// * `mem` - the `GuestMemory` object that can be used to access the queue buffers.
+    fn iter<M>(&mut self, mem: M) -> Result<AvailIter<'_, M>, Error>
+    where
+        M: Deref,
+        M::Target: GuestMemory;
+
+    /// Undo the last advancement of the next available index field by decrementing its
+    /// value by one.
+    fn go_to_previous_position(&mut self);
```

Comment:
This is still design flaw about `go_to_previous_position()`. There is no clear definition
whether caller could call `go_to_previous_position()` multiple times and what's the expected behavior.
All that we have is `all callers will only call it once`.

Reply:
This is an interesting and tricky to improve situation. Add the method here does not enable users of the crate to do anything that wasn't already possible before with `iter().go_to_previous_position()`. In terms of multiple consumers concurrently calling the method and thus ending in an undefined situation, this should not happen because the trait stands for queues that are handled by a single entity. You are right that even a single consumer can theoretically call this a sufficiently large number of times and break things, but it's not super clear right now how that can be avoided without making things difficult to use (also, `add_used` is in a similar situation, as it can be called any number of times with the same or invalid indices).

One approach can be to require the consumption of a `DescriptorChain` (that cannot be cloned, or manually instantiated, etc.) for these operations (but even that wouldn't be absolutely foolproof I think, because we can get a chain from another queue for example). We need to think a bit more about what a better solution can be here, but meanwhile I think the existing methods are serviceable and not that prone to misuse.

Reply:
The flaw is not introduced by this PR, actually it already exists for a long time. So let's handle it separately.